src/Compilers/Core/Portable/DiagnosticAnalyzer/SymbolDeclaredCompilationEvent.cs

```diff
@@ -16,83 +18,26 @@ internal sealed class SymbolDeclaredCompilationEvent : CompilationEvent
     {
         private readonly Lazy<ImmutableArray<SyntaxReference>> _lazyCachedDeclaringReferences;
 
-        public SymbolDeclaredCompilationEvent(Compilation compilation, ISymbol symbol) : base(compilation)
+        public SymbolDeclaredCompilationEvent(Compilation compilation, ISymbol symbol, SemanticModel? semanticModelWithCachedBoundNodes = null)
+            : base(compilation)
         {
-            this.Symbol = symbol;
-            this._lazyCachedDeclaringReferences = new Lazy<ImmutableArray<SyntaxReference>>(() => symbol.DeclaringSyntaxReferences);
-        }
-
-        public SymbolDeclaredCompilationEvent(Compilation compilation, ISymbol symbol, Lazy<SemanticModel> lazySemanticModel) : this(compilation, symbol)
-        {
-            _lazySemanticModel = lazySemanticModel;
-        }
-
-        private SymbolDeclaredCompilationEvent(SymbolDeclaredCompilationEvent original, SemanticModel newSemanticModel) : this(original.Compilation, original.Symbol)
-        {
-            _semanticModel = newSemanticModel;
+            Symbol = symbol;
+            SemanticModelWithCachedBoundNodes = semanticModelWithCachedBoundNodes;
+            _lazyCachedDeclaringReferences = new Lazy<ImmutableArray<SyntaxReference>>(() => symbol.DeclaringSyntaxReferences);
```

Comment:
>_lazyCachedDeclaringReferences [](start = 12, length = 30)

Consider making this a lazily-initialized ImmutableArray with `ImmutableInterlocked.Initialize`, instead of using a `Lazy<T>`.

Reply:
This is existing code, so I was trying to not change it in this PR.

Reply:
Consider either doing it inline, or filling an issue to follow up. We already do perf-sensitive work in this class, and this is a very-commonly-allocated class type.

---
In reply to: [469652219](https://github.com/dotnet/roslyn/pull/46508#discussion_r469652219) [](ancestors = 469652219)